src/meta/meta_ckpt.c

```diff
@@ -503,42 +537,211 @@ __wt_meta_blk_mods_load(WT_SESSION_IMPL *session, const char *config, WT_CKPT *c
  */
 int
 __wt_meta_ckptlist_get(
-  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep)
+  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep, size_t *allocated)
 {
     WT_DECL_RET;
     char *config;
 
     config = NULL;
 
     WT_ERR(__wt_metadata_search(session, fname, &config));
-    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, config));
+    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, allocated, config));
 
 err:
     __wt_free(session, config);
     return (ret);
 }
 
+#ifdef HAVE_DIAGNOSTIC
+/*
+ * __assert_ckpt_matches --
+ *     Assert that given two checkpoints match.
+ */
+static void
+__assert_ckpt_matches(WT_SESSION_IMPL *session, WT_CKPT *ckpt_a, WT_CKPT *ckpt_b)
+{
+    /*
+     * We are not checking checkpoint time, because there could be a minute difference depending
+     * upon when the checkpoint information was generated. This is acceptable.
+     */
+    WT_ASSERT(session,
+      (ckpt_a->name == NULL && ckpt_b->name == NULL) ||
+        (ckpt_a->name != NULL && ckpt_b->name != NULL && strcmp(ckpt_a->name, ckpt_b->name) == 0));
+    WT_ASSERT(session, ckpt_a->order == ckpt_b->order);
+    WT_ASSERT(session, ckpt_a->size == ckpt_b->size);
+    WT_ASSERT(session, ckpt_a->write_gen == ckpt_b->write_gen);
+    WT_ASSERT(session, ckpt_a->run_write_gen == ckpt_b->run_write_gen);
+    WT_ASSERT(session,
+      ckpt_a->ta.newest_start_durable_ts == ckpt_b->ta.newest_start_durable_ts &&
+        ckpt_a->ta.newest_stop_durable_ts == ckpt_b->ta.newest_stop_durable_ts &&
+        ckpt_a->ta.oldest_start_ts == ckpt_b->ta.oldest_start_ts &&
+        ckpt_a->ta.newest_txn == ckpt_b->ta.newest_txn &&
+        ckpt_a->ta.newest_stop_ts == ckpt_b->ta.newest_stop_ts &&
+        ckpt_a->ta.newest_stop_txn == ckpt_b->ta.newest_stop_txn &&
+        ckpt_a->ta.prepare == ckpt_b->ta.prepare &&
+        __wt_check_addr_validity(session, &ckpt_a->ta, true) == 0);
+    // WT_ASSERT(session, compare ckpt_a->addr.data, ckpt_b->addr.data, ckpt_a->addr.size == 0);
+    WT_ASSERT(session, ckpt_a->bpriv == NULL && ckpt_b->bpriv == NULL);
+    WT_ASSERT(session, ckpt_a->flags == ckpt_b->flags);
+}
+
+/*
+ * __assert_checkpoint_list_matches --
+ *     Assert that two given checkpoint lists match.
+ */
+static void
+__assert_checkpoint_list_matches(WT_SESSION_IMPL *session, WT_CKPT *saved_list, WT_CKPT *new_list)
+{
+    WT_CKPT *ckpt_saved, *ckpt_new;
+
+    for (ckpt_saved = saved_list, ckpt_new = new_list;
+         ckpt_saved != NULL && ckpt_saved->order != 0 && ckpt_new != NULL && ckpt_new->order != 0;
+         ckpt_saved++, ckpt_new++)
+        __assert_ckpt_matches(session, ckpt_saved, ckpt_new);
+
+    if (!((ckpt_saved == NULL && ckpt_new == NULL) ||
+          ((ckpt_saved != NULL && ckpt_saved->order == 0) &&
+            (ckpt_new != NULL && ckpt_new->order == 0))))
+        WT_ASSERT(session, false);
+}
```

Comment:
I would write this as
```
    WT_ASSERT(session, ckpt_saved == NULL && ckpt_new == NULL || 
          ((ckpt_saved != NULL && ckpt_saved->order == 0) &&
            (ckpt_new != NULL && ckpt_new->order == 0)));
```
I think it's simpler and more intuitive.

Would it be simpler to break out when you find a WT_CKPT structure with `WT_CKPT_ADD` instead of looking for `order == 0`? I mean, isn't `order == 0` an error if you find it in the middle?

Reply:
That is my weird style of writing asserts because I replace it with printfs when debugging. It must have slipped past me when sending it out for final review. I have fixed the assert.

> Would it be simpler to break out when you find a WT_CKPT structure with WT_CKPT_ADD instead of looking for order == 0? I mean, isn't order == 0 an error if you find it in the middle?
I was trying to be generic with the checkpoint-list-matches function, that would work even if there was no checkpoint getting added at the end. I think this followed from how I understood the code as I wrote this change. Ideally, there would be a non-refutable means to tell the end of the list, but absent that I figured I have to rely on `ckpt->order == 0`. 

If there is not a strong objection, I will like to keep the assert as you suggested, and break out when order is 0.